libyul/ASTLabelRegistry.cpp

```diff
@@ -0,0 +1,194 @@
+#include <range/v3/view/map.hpp>
+
+using namespace solidity::yul;
+
+namespace
+{
+
+bool isInGhostFormat(std::string_view const _label)
+{
+	using namespace std::literals::string_view_literals;
+	static size_t constexpr prefixLength = "GHOST["sv.length();
+	static size_t constexpr postfixLength = "]"sv.length();
+	if (_label.size() < prefixLength + postfixLength + 1)
+		return false;
+	if (!_label.starts_with("GHOST[") || !_label.ends_with("]"))
+		return false;
+
+	auto const maybeID = _label.substr(prefixLength, _label.size() - prefixLength - postfixLength);
+	return ranges::all_of(maybeID, isdigit);
+}
+
+}
+
+ASTLabelRegistry::ASTLabelRegistry(): m_labels{""}, m_idToLabelMapping{0} {}
+
+ASTLabelRegistry::ASTLabelRegistry(std::vector<std::string> _labels, std::vector<size_t> _idToLabelMapping)
+{
+	yulAssert(!_labels.empty());
+	yulAssert(_labels[0].empty());
+	yulAssert(!_idToLabelMapping.empty());
+	yulAssert(_idToLabelMapping[0] == 0);
+	// using vector<uint8_t> over vector<bool>, as the latter is optimized for space-efficiency
+	std::vector<uint8_t> labelVisited (_labels.size(), false);
+	size_t numLabels = 0;
+	for (auto const& labelIndex: _idToLabelMapping)
+	{
+		if (labelIndex == ghostID())
+			continue;
+		yulAssert(labelIndex < _labels.size());
+		// it is possible to have multiple references to empty / ghost
+		yulAssert(
+			labelIndex == 0 || !labelVisited[labelIndex],
+			fmt::format("LabelID {} (label \"{}\") is not unique.", labelIndex, _labels[labelIndex])
+		);
+		labelVisited[labelIndex] = true;
+		yulAssert(!isInGhostFormat(_labels[labelIndex]), "Labels of the form GHOST[id] are reserved.");
+		if (labelIndex >= 1)
+			++numLabels;
+	}
+	yulAssert(numLabels + 1 == _labels.size(), "Unused labels present.");
+	m_labels = std::move(_labels);
+	m_idToLabelMapping = std::move(_idToLabelMapping);
+}
+
+ASTLabelRegistry::LabelID ASTLabelRegistry::maximumID() const
+{
+	yulAssert(!m_idToLabelMapping.empty());
+	return m_idToLabelMapping.size() - 1;
+}
+
+size_t ASTLabelRegistry::idToLabelIndex(LabelID const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size());
+	return m_idToLabelMapping[_id];
+}
+
+std::string_view ASTLabelRegistry::operator[](LabelID const _id) const
+{
+	auto const labelIndex = idToLabelIndex(_id);
+	if (labelIndex == ghostID())
+		return lookupGhost(_id);
+	return m_labels[labelIndex];
+}
+
+std::optional<ASTLabelRegistry::LabelID> ASTLabelRegistry::findIdForLabel(std::string_view const _label) const {
+	if (_label.empty())
+		return emptyID();
+	for (LabelID id = 1; id <= maximumID(); ++id)
+		if ((*this)[id] == _label)
+			return id;
+	return std::nullopt;
+}
+
+std::string_view ASTLabelRegistry::lookupGhost(LabelID const _id) const
+{
+	yulAssert(idToLabelIndex(_id) == ghostID());
+	auto const [it, _] = m_ghostLabelCache.try_emplace(_id, fmt::format("GHOST[{}]", _id));
+	return it->second;
+}
+
+ASTLabelRegistryBuilder::DefinedLabels::DefinedLabels():
+	m_mapping{{"", 0}}
```

Comment:
```suggestion
	m_labelToIDMapping{{"", 0}}
```

Also, you could just do `m_labelToIDMapping = {{"", 0}};` in class body.